### PR TITLE
fix: [L-09] restrict owner change in _beforeTokenTransfer

### PR DESCRIPTION
**Pull request by skimaharvey**
# What does this PR introduce? 

- [x] Restrict owner change in _beforeTokenTransfer

### CODE REVIEW

File: contracts/LSP8IdentifiableDigitalAsset/LSP8Errors.sol [hunk 1]
```diff
@@ -103,3 +103,8 @@ error LSP8TokenIdsDataEmptyArray();
  * @notice Batch call failed.
  */
 error LSP8BatchCallFailed(uint256 callIndex);
+
+/**
+ * @dev Reverts when the token owner has changed after the _beforeTokenTransfer hook.
```
Comment (CJ42) [inline contracts/LSP8IdentifiableDigitalAsset/LSP8Errors.sol:108]: Let's make that as an anchor link to the internal function for the docs:

```suggestion
 * @dev Reverts when the token owner changed inside the {_beforeTokenTransfer} hook.
```

File: contracts/Mocks/Tokens/LSP8TransferOwnerChange.sol [hunk 2]
```diff
@@ -0,0 +1,51 @@
...
+    }
+
+    function _beforeTokenTransfer(
+        address,
+        address,
+        bytes32 tokenId,
+        bytes memory
+    ) internal override {
+        // if tokenID exist transfer token ownership to this contract
+        if (_tokenOwners[tokenId] != address(0)) {
```
Comment (CJ42) [inline contracts/Mocks/Tokens/LSP8TransferOwnerChange.sol:47]: Maybe we can use the internal function here?

```suggestion
        if (_exists(tokenId)) {
```

image

File: contracts/LSP8IdentifiableDigitalAsset/LSP8Errors.sol [hunk 3]
```diff
@@ -103,3 +103,8 @@ error LSP8TokenIdsDataEmptyArray();
...
+ */
+error LSP8TokenOwnerChanged(address oldOwner, address newOwner);
```
Comment (CJ42) [inline contracts/LSP8IdentifiableDigitalAsset/LSP8Errors.sol:110]: I think we should also specify for which `tokenId` the owner has changed here. Wdyt?

```suggestion
error LSP8TokenIdOwnerChanged(bytes32 tokenId, address oldOwner, address newOwner);
```

File: contracts/LSP8IdentifiableDigitalAsset/LSP8IdentifiableDigitalAssetCore.sol [hunk 4]
```diff
@@ -634,9 +635,11 @@ abstract contract LSP8IdentifiableDigitalAssetCore is
 
         _beforeTokenTransfer(from, to, tokenId, data);
 
-        // Re-fetch and update `tokenOwner` in case `tokenId`
-        // was transferred inside the `_beforeTokenTransfer` hook
-        tokenOwner = tokenOwnerOf(tokenId);
+        // Check that `tokenId`'s owner was not changed inside the `_beforeTokenTransfer` hook
+        address currentTokenOwner = tokenOwnerOf(tokenId);
+        if (tokenOwner != currentTokenOwner) {
+            revert LSP8TokenOwnerChanged(tokenOwner, currentTokenOwner);
```
Comment (CJ42) [inline contracts/LSP8IdentifiableDigitalAsset/LSP8IdentifiableDigitalAssetCore.sol:641]: Suggest using named parameters for readability here:

```suggestion
            revert LSP8TokenOwnerChanged({
                oldOwner: tokenOwner, 
                newOwner: currentTokenOwner
            });
```

**Review by CJ42 — COMMENTED** [review]
Added review comments

File: contracts/LSP8IdentifiableDigitalAsset/LSP8IdentifiableDigitalAssetCore.sol (hunk 4 above)
Comment (skimaharvey) [reply contracts/LSP8IdentifiableDigitalAsset/LSP8IdentifiableDigitalAssetCore.sol:641]: well, in most of our errors we don't use them. If I do for this one we will need to update all the errors out of consistency

Comment (CJ42) [reply contracts/LSP8IdentifiableDigitalAsset/LSP8IdentifiableDigitalAssetCore.sol]: Ok let's keep it as it is for this one.

**Review by CJ42 — APPROVED** [review]

**Review by b00ste — APPROVED** [review]

**Pull request merged**